core/node/registries/river_registry_contract.go

```diff
@@ -393,13 +423,133 @@ func (c *RiverRegistryContract) ForAllStreams(
 			if err != nil {
 				return err
 			}
+			totalStreams++
 			if !cb(makeGetStreamResult(streamId, &stream.Stream)) {
 				return nil
 			}
 		}
 	}
 
-	log.Info("RiverRegistryContract: GetPaginatedStreams completed", "elapsed", time.Since(startTime))
+	elapsed := time.Since(startTime)
+	log.Info(
+		"RiverRegistryContract: GetPaginatedStreams completed",
+		"elapsed",
+		elapsed,
+		"streamsPerSecond",
+		float64(totalStreams)/elapsed.Seconds(),
+	)
+
+	return nil
+}
+
+// ForAllStreams calls the given cb for all streams that are registered in the river registry at the given block num.
+// If cb returns false forAllStreams returns.
+func (c *RiverRegistryContract) forAllStreamsParallel(
+	ctx context.Context,
+	blockNum crypto.BlockNumber,
+	cb func(*GetStreamResult) bool,
+) error {
+	log := dlog.FromCtx(ctx)
+
+	numWorkers := c.Settings.ParallelReaders
+	if numWorkers <= 1 {
+		numWorkers = 8
+	}
+
+	pageSize := int64(c.Settings.PageSize)
+	if pageSize <= 0 {
+		pageSize = 1000
+	}
+
+	progressReportInterval := c.Settings.ProgressReportInterval
+	if progressReportInterval <= 0 {
+		progressReportInterval = 10 * time.Second
+	}
+
+	numStreamsBigInt, err := c.StreamRegistry.GetStreamCount(c.callOptsWithBlockNum(ctx, blockNum))
+	if err != nil {
+		return WrapRiverError(Err_CANNOT_CALL_CONTRACT, err).Func("ForAllStreams")
+	}
+
+	chResults := make(chan []river.StreamWithId, numWorkers)
+	chErrors := make(chan error, numWorkers)
+
+	pool := workerpool.New(numWorkers)
+	defer pool.StopWait()
+
+	startTime := time.Now()
+	lastReport := time.Now()
+
+	numStreams := numStreamsBigInt.Int64()
+	taskCounter := atomic.NewInt64(0)
+	for i := int64(0); i < numStreams; i += pageSize {
+		taskCounter.Inc()
+		pool.Submit(func() {
+			streams, _, err := c.callGetPaginatedStreamsWithBackoff(ctx, blockNum, i, i+pageSize)
+			if err == nil {
+				chResults <- streams
+			} else {
+				chErrors <- err
+			}
+			taskCounter.Dec()
+		})
+	}
+
+	totalStreams := int64(0)
+OuterLoop:
+	for {
+		now := time.Now()
+		if now.Sub(lastReport) > progressReportInterval {
+			elapsed := time.Since(startTime)
+			log.Info(
+				"RiverRegistryContract: GetPaginatedStreams in progress",
+				"streamsRead",
+				totalStreams,
+				"elapsed",
+				elapsed,
+				"streamPerSecond",
+				float64(totalStreams)/elapsed.Seconds(),
+			)
+			lastReport = now
+		}
+		select {
+		case streams := <-chResults:
+			for _, stream := range streams {
+				if stream.Id == ZeroBytes32 {
+					continue
+				}
+				totalStreams++
+				if !cb(makeGetStreamResult(stream.Id, &stream.Stream)) {
+					break OuterLoop
+				}
+			}
+			if taskCounter.Load() == 0 {
+				break OuterLoop
+			}
+		case receivedErr := <-chErrors:
+			err = receivedErr
+			break OuterLoop
+		case <-ctx.Done():
```

Comment:
Should we set a reasonable ctx timeout when calling parallel stream get ? If not, then there's the potential for a deadlock if chErrors, chResults get stuck where this just hangs.

Reply:
yes, it sets timeout and number of retries from settings